pkg/collector/corechecks/cluster/kubernetes_state_label_joins.go

```diff
@@ -0,0 +1,208 @@
+     kube_deployment_labels {"namespace": "ns-a", "deployment": "bar", "chart_name": "bar", "chart_version": "1.1"}
+     kube_deployment_labels {"namespace": "ns-b", "deployment": "foo", "chart_name": "foo", "chart_version": "2.0"}
+
+   it will create the following tree:
+
+   kube_deployment_labels
+     ├─ ns-a
+     │    ├─ foo
+     │    │    └─ [ chart_name:foo, chart_version:1.0 ]
+     │    └─ bar
+     │         └─ [ chart_name:bar, chart_version:1.1 ]
+     └─ ns-b
+          └─ foo
+               └─ [ chart_name:foo, chart_version:2.0 ]
+
+   At the first level of the tree, there are the different values for the "namespace" label (because it’s the first label to match)
+   At the second level of the tree, there are all the different values for the "deployment" label (because it’s the first label to match)
+   At the third level of the tree, there are the lists of labels to add with the keys and the values.
+
+   When a metric like the following needs to be decorated:
+
+     kube_pod_container_status_running {"namespace": "ns-a", "deployment": "bar", "container": "agent", "pod": "XXX"}
+
+   We first extract the "namespace" value because it’s the first label to match.
+   This value is used to lookup the first level node in the tree.
+   The "deployment" value is then extracted because it’s the second label to match.
+   This value is used to lookup the second level node in the tree.
+   That node contains the list of labels to add.
+*/
+
+type labelJoiner struct {
+	metricsToJoin map[string]metricToJoin
+}
+
+type metricToJoin struct {
+	config *JoinsConfig
+	tree   *node
+}
+
+type label struct {
+	key   string
+	value string
+}
+
+type node struct {
+	labelValues map[string]*node
+	labelsToAdd []label
+}
+
+func newLabelJoiner(config map[string]*JoinsConfig) *labelJoiner {
+	metricsToJoin := make(map[string]metricToJoin)
+
+	for key, joinsConfig := range config {
+		if len(joinsConfig.LabelsToMatch) > 0 {
+			metricsToJoin[key] = metricToJoin{
+				config: joinsConfig,
+				tree:   newInnerNode(),
+			}
+		} else {
+			metricsToJoin[key] = metricToJoin{
+				config: joinsConfig,
+				tree:   newLeafNode(),
+			}
+		}
+	}
+
+	return &labelJoiner{
+		metricsToJoin: metricsToJoin,
+	}
+}
+
+func newInnerNode() *node {
+	return &node{
+		labelValues: make(map[string]*node),
+		labelsToAdd: nil,
+	}
+}
+
+func newLeafNode() *node {
+	return &node{
+		labelValues: nil,
+		labelsToAdd: nil,
+	}
+}
+
+func (lj *labelJoiner) insertMetric(metric ksmstore.DDMetric, config *JoinsConfig, tree *node) {
+	current := tree
+	nbLabelsToMatch := len(config.LabelsToMatch)
+	for i, labelToMatch := range config.LabelsToMatch {
+		labelValue, found := metric.Labels[labelToMatch]
+		if !found {
+			return
+		}
+
+		child, found := current.labelValues[labelValue]
+		if !found {
+			if i < nbLabelsToMatch-1 {
+				child = newInnerNode()
+			} else {
+				child = newLeafNode()
+			}
```

Comment:
mind adding some comments here to explain what we're doing?

Reply:
Just added some comments in this function.